numpy/_core/src/umath/stringdtype_ufuncs.cpp

```diff
@@ -1881,6 +1881,167 @@ zfill_strided_loop(PyArrayMethod_Context *context,
+                     "support the 'out' keyword", self->name);
+        return (NPY_CASTING)-1;
+    }
+    for (int i=0; i<3; i++) {
+        Py_INCREF(given_descrs[i]);
+        loop_descrs[i] = given_descrs[i];
+    }
+    PyArray_StringDTypeObject *adescr = (PyArray_StringDTypeObject *)given_descrs[0];
+    for (int i=3; i<6; i++) {
+        loop_descrs[i] = (PyArray_Descr *)new_stringdtype_instance(
+                adescr->na_object, adescr->coerce);
+        if (loop_descrs[i] == NULL) {
+            return (NPY_CASTING)-1;
+        }
+    }
+
+    return NPY_NO_CASTING;
+}
+
+NPY_NO_EXPORT int
+string_partition_strided_loop(
+        PyArrayMethod_Context *context,
+        char *const data[],
+        npy_intp const dimensions[],
+        npy_intp const strides[],
+        NpyAuxData *NPY_UNUSED(auxdata))
+{
+    STARTPOSITION startposition = *(STARTPOSITION *)(context->method->static_data);
+
+    npy_intp N = dimensions[0];
+
+    char *in1 = data[0];
+    char *in2 = data[1];
+    char *in3 = data[2];
+    char *out1 = data[3];
+    char *out2 = data[4];
+    char *out3 = data[5];
+
+    npy_intp in1_stride = strides[0];
+    npy_intp in2_stride = strides[1];
+    npy_intp in3_stride = strides[2];
+    npy_intp out1_stride = strides[3];
+    npy_intp out2_stride = strides[4];
+    npy_intp out3_stride = strides[5];
+
+    npy_string_allocator *allocators[6] = {};
+    NpyString_acquire_allocators(6, context->descriptors, allocators);
+    npy_string_allocator *in1allocator = allocators[0];
+    npy_string_allocator *in2allocator = allocators[1];
+    // allocators[2] is NULL
+    npy_string_allocator *out1allocator = allocators[3];
+    npy_string_allocator *out2allocator = allocators[4];
+    npy_string_allocator *out3allocator = allocators[5];
+
+    PyArray_StringDTypeObject *idescr =
+            (PyArray_StringDTypeObject *)context->descriptors[0];
+    int has_string_na = idescr->has_string_na;
+    const npy_static_string *default_string = &idescr->default_string;
+
+    while (N--) {
+        const npy_packed_static_string *i1ps = (npy_packed_static_string *)in1;
+        npy_static_string i1s = {0, NULL};
+        const npy_packed_static_string *i2ps = (npy_packed_static_string *)in2;
+        npy_static_string i2s = {0, NULL};
+        npy_packed_static_string *o1ps = (npy_packed_static_string *)out1;
+        npy_packed_static_string *o2ps = (npy_packed_static_string *)out2;
+        npy_packed_static_string *o3ps = (npy_packed_static_string *)out3;
+
+        int i1_isnull = NpyString_load(in1allocator, i1ps, &i1s);
+        int i2_isnull = NpyString_load(in2allocator, i2ps, &i2s);
+
+        if (i1_isnull == -1 || i2_isnull == -1) {
+            npy_gil_error(PyExc_MemoryError, "Failed to load string in %s",
+                          ((PyUFuncObject *)context->caller)->name);
+            goto fail;
+        }
+        else if (NPY_UNLIKELY(i1_isnull || i2_isnull)) {
+            if (!has_string_na) {
```

Comment:
I decided against adding support for nan-like nulls. Happy to add it if reviewers think it's needed.

Reply:
I guess one has to be slightly careful, since one wants `sum(parts)` to work, so only the left or right result should be `NaN`, the others still empty strings. It seems that would work automatically, since `find` would already not return a result.